### ISSUE TITLE
No subsurface keyboard focus

### ISSUE THREAD

**Issue by nyorain** [swaywm/sway#5904]
With sway version 1.5.1:
Subsurfaces don't get keyboard focus. When creating a subsurface for an xdg_toplevel surface and clicking on the subsurface, the wl_keyboard does not change focus (and therefore future key events cannot be delievered to the subsurface as expected). It works as expected for wl_pointer.{enter, leave}. On weston, the keyboard does change focus to the subsurface.

The wayland spec does not explicitly state that subsurfaces can receive focus but I did expect it and since weston does it that way, other applications might expect it as well. Would you consider this a bug and accept a fix for it (not sure I'll find the time but might look into it if no one else does so first) or is this more of a design decision?

**Comment by emersion** [comment]
>since weston does it that way, other applications might expect it as well

That's not a good reason why we should change our behavior.

>Would you consider this a bug

If the Wayland protocol doesn't state that we should, it's probably not a Sway bug.

>accept a fix for it

Maybe, if it doesn't increase code complexity. I don't know what the most sensible behavior is.

Do you have an example of a client being broken because of this?

**Comment by nyorain** [comment]
Yes, strictly speaking not a bug. But I'd argue that subsurfaces being able to get keyboard focus is expected behavior as that's how it works in all window systems (and wayland even offers input_region to exclude surfaces from input, if desired). I don't have a specific client that breaks but am porting a client from x11 that uses child windows to block input from reaching the main window and instead handling it in the child window. If you disagree on expected behavior feel free to just close the issue but I really feel like this shouldn't be up to compositors to decide as it makes subsurfaces useless regarding input. Maybe it should be clarified in the wayland protocol documentation?

**Comment by emersion** [comment]
Yeah, some protocol clarifications would be good.

**Comment by axelkar** [comment]
Hi. Is there progress on this? How do Mutter and KWin implement this?

**Comment by emersion** [comment]
Per [wayland!236](https://gitlab.freedesktop.org/wayland/wayland/-/merge_requests/236), sub-surfaces are never to receive keyboard focus.
